tests/MongoDB.Driver.Tests/Specifications/unified-test-format/UnifiedTestOperations/UnifiedBulkWriteOperation.cs

```diff
@@ -0,0 +1,274 @@
+
+            return new ReplaceOneModel<BsonDocument>(filter, replacement)
+            {
+                IsUpsert = isUpsert
+            };
+        }
+
+        private UpdateManyModel<BsonDocument> ParseUpdateManyModel(BsonDocument model)
+        {
+            JsonDrivenHelper.EnsureAllFieldsAreValid(model, "filter", "update", "upsert");
+
+            var filter = new BsonDocumentFilterDefinition<BsonDocument>(model["filter"].AsBsonDocument);
+            var update = new BsonDocumentUpdateDefinition<BsonDocument>(model["update"].AsBsonDocument);
+            var isUpsert = model.GetValue("upsert", false).ToBoolean();
+
+            return new UpdateManyModel<BsonDocument>(filter, update)
+            {
+                IsUpsert = isUpsert
+            };
+        }
+
+        private UpdateOneModel<BsonDocument> ParseUpdateOneModel(BsonDocument model)
+        {
+            JsonDrivenHelper.EnsureAllFieldsAreValid(model, "filter", "update", "upsert");
+
+            var filter = new BsonDocumentFilterDefinition<BsonDocument>(model["filter"].AsBsonDocument);
+            var update = new BsonDocumentUpdateDefinition<BsonDocument>(model["update"].AsBsonDocument);
+            var isUpsert = model.GetValue("upsert", false).ToBoolean();
+
+            return new UpdateOneModel<BsonDocument>(filter, update)
+            {
+                IsUpsert = isUpsert
+            };
+        }
+
+        private WriteModel<BsonDocument> ParseWriteModel(BsonDocument modelItem)
+        {
+            if (modelItem.ElementCount > 1)
+            {
+                throw new FormatException("BulkWrite request model should contain single element");
+            }
+
+            var modelName = modelItem.GetElement(0).Name;
+            var model = modelItem[0].AsBsonDocument;
+            switch (modelName)
+            {
+                case "deleteMany":
+                    return ParseDeleteManyModel(model);
+                case "deleteOne":
+                    return ParseDeleteOneModel(model);
+                case "insertOne":
+                    return ParseInsertOneModel(model);
+                case "replaceOne":
+                    return ParseReplaceOneModel(model);
+                case "updateMany":
+                    return ParseUpdateManyModel(model);
+                case "updateOne":
+                    return ParseUpdateOneModel(model);
+                default:
+                    throw new FormatException($"Invalid write model name: {modelName}.");
+            }
+        }
+
+        private List<WriteModel<BsonDocument>> ParseWriteModels(IEnumerable<BsonDocument> models)
+        {
+            var result = new List<WriteModel<BsonDocument>>();
+            foreach (var model in models)
+            {
+                result.Add(ParseWriteModel(model));
+            }
+
+            return result;
+        }
+    }
+
+    public class UnifiedBulkWriteOperationResultConverter
```

Comment:
In cases when result is not trivially converted to bson I think it is better to move the functionality out of the operation class.

Reply:
Marginal benefit. But OK.